### PR TITLE
Project Creation/Layout

### PR DESCRIPTION
**Pull request by caseymm**
#### What is this change?
Reworked format for creating projects and updated style to resemble Anthem.

#### Why is it being made?
To better enable users to make projects in a more seamless fashion in a way that feels intuitive and mirrors the process of creating an article in Anthem as closely as possible.

#### What are the risks?
I think the biggest risk is that there are a lot of `if/else` statements now regarding live preview projects vs graphics vs apps. I _think_ I caught them all and I tried to test everything, but it's possible that I could have missed something.

### CODE REVIEW

File: appjs/views/edit_project.js [hunk 1]
```diff
@@ -153,6 +154,45 @@ var EditProject = BaseView.extend(require('./mixins/actions'), require('./mixins
...
+          $('#preview-pane').css("width", event.pageX);
+          view.showPreviewButtons();
+        }
+      });
+    }
+
+    $(document).mouseup(function (e) {
+      $('#embed-preview').removeClass('screen');
+      $(document).unbind('mousemove');
+    });
```
Comment (caseymm) [inline appjs/views/edit_project.js]: Allows for user to pull/drag form in and out. Helpful when creating projects like the quiz where narrow forms can be a hassle.

File: appjs/views/edit_project.js [hunk 2]
```diff
@@ -153,6 +154,45 @@ var EditProject = BaseView.extend(require('./mixins/actions'), require('./mixins
...
+      $('.nav-pills #medium-view').hide();
+    } else {
+      $('.nav-pills .resize#small-view').trigger('click');
+      $('.nav-pills button').hide();
+      if($(window).width() > 768){
+        $('.nav-pills .resize#small-view').show();
+      }
+    }
+    $('.nav-pills li button').show();
+  },
```
Comment (caseymm) [inline appjs/views/edit_project.js]: Show/hide preview buttons depending on the current width of the preview frame.

File: appjs/views/edit_project.js [hunk 3]
```diff
@@ -170,85 +210,113 @@ var EditProject = BaseView.extend(require('./mixins/actions'), require('./mixins
   },
 
   pollChange: _.debounce(function(){
+    this.alpaca.childrenByPropertyId["tweet_text"].setValue($('textarea#shareText').val());
```
Comment (caseymm) [inline appjs/views/edit_project.js]: Sets the value of the hidden `tweet_text` field in the alpaca form to that of the `#shareText` element on the `publish` tab.

File: appjs/views/edit_project.js [hunk 4]
```diff
@@ -170,85 +210,113 @@ var EditProject = BaseView.extend(require('./mixins/actions'), require('./mixins
...
+          view.app.view.error(
+            "Could not update live preview, please contact support.", 'permanent' );
+          logger.error(err);
+        }
+      });
+    }
+  }, 500),
 
-        if ( view.pym ) { view.pym.remove(); }
+  getTwitterCount: function(){
```
Comment (caseymm) [inline appjs/views/edit_project.js]: Sets the available character count for share text.
screen shot 2016-07-05 at 4 43 04 pm

File: appjs/views/edit_project.js [hunk 5]
```diff
@@ -325,6 +393,19 @@ var EditProject = BaseView.extend(require('./mixins/actions'), require('./mixins
...
+      if(view.formWidth){
+        if($(window).width() > 768){
+          $('#form-pane').css("width", view.formWidth);
+          $('#preview-pane').css("width", $(window).width() - view.formWidth);
+        } else {
+          $('#form-pane').css("width", '100%');
+          $('#preview-pane').css("width", '100%');
+        }
+      }
+    });
```
Comment (caseymm) [inline appjs/views/edit_project.js]: Retain the width of the form panel that the user has set even if the window is resized.

File: appjs/views/edit_project.js [hunk 6]
```diff
@@ -364,6 +445,7 @@ var EditProject = BaseView.extend(require('./mixins/actions'), require('./mixins
             previewUrl = '', iframeLoaded,
             previewSlug = '';
 
+        view.$('#shareText').val(formData['tweet_text']);
```
Comment (caseymm) [inline appjs/views/edit_project.js]: Set the value of the `#shareText` element within the publish tab to contain the share_text that has been saved in the alpaca form after the page is rendered.

File: appjs/views/edit_project.js [hunk 7]
```diff
@@ -404,16 +483,19 @@ var EditProject = BaseView.extend(require('./mixins/actions'), require('./mixins
         if ( view.model.hasType( 'graphic' ) || view.model.hasPreviewType('live') ) {
           // Setup our iframe with pym
           if ( view.pym ) { view.pym.remove(); }
-          view.pym = new pym.Parent('embed-preview', previewUrl);
-          view.pym.iframe.onload = iframeLoaded;
-
+          if ( view.formValidate(view.model, view.$('#projectForm')) ){
+            view.pym = new pym.Parent('embed-preview', previewUrl);
+            view.pym.iframe.onload = iframeLoaded;
+          }
```
Comment (caseymm) [inline appjs/views/edit_project.js]: Only load the preview if the form is valid. Prevents the project preview from loading until all/enough information has been entered in order to generate a correct preview.

File: appjs/views/edit_project.js [hunk 8]
```diff
@@ -404,16 +483,19 @@ var EditProject = BaseView.extend(require('./mixins/actions'), require('./mixins
...
           // In case some dumb script hangs the loading process
           setTimeout(iframeLoaded, 20000);
-
-          if ( view.togglePreview ){
-            $( "#draft-preview" ).trigger( "click" );
+        }
+        if(view.model.hasPreviewType('live')){
+          if(view.model.getConfig().spreadsheet_template){
+            $( "input[name='google_doc_url']" ).after('<button type="button" id="spreadsheet-button" data-hook="create-spreadsheet" class="btn btn-default">Get new spreadsheet</button>');
```
Comment (caseymm) [inline appjs/views/edit_project.js]: Position the `Get new spreadsheet` button to appear below the spreadsheet input box as opposed to being located above it.

File: appjs/views/edit_project.js [hunk 9]
```diff
@@ -404,16 +483,19 @@ var EditProject = BaseView.extend(require('./mixins/actions'), require('./mixins
...
           }
         }
+        view.getTwitterCount();
```
Comment (caseymm) [inline appjs/views/edit_project.js]: Get the twitter count for `#shareText` on project render so that it appears automatically and doesn't require the user to type before showing the remaining character count.

File: appjs/views/edit_project.js [hunk 10]
```diff
@@ -466,7 +547,7 @@ var EditProject = BaseView.extend(require('./mixins/actions'), require('./mixins
         return _.contains(this.model.getConfig().themes, t.get('slug'));
       }, this)) : this.app.themes.models;
     availableThemes = availableThemes || this.app.themes.where({slug : 'generic'});
-    twitterHandles = _.object(this.app.themes.pluck('slug'), this.app.themes.pluck('twitter_handle'));
+    this.twitterHandles = _.object(this.app.themes.pluck('slug'), this.app.themes.pluck('twitter_handle'));
```
Comment (caseymm) [inline appjs/views/edit_project.js]: Saved variable as part of `this` so that it can be accessed from `getTwitterCount` (which runs on `keyup` from `#shareText`) and not generated every time.

File: appjs/views/edit_project.js [hunk 11]
```diff
@@ -565,7 +641,13 @@ var EditProject = BaseView.extend(require('./mixins/actions'), require('./mixins
...
       var opts = {
         "schema": {
-          "title": this.model.blueprint.get('title'),
+          "title": function(){
+            if(view.model.hasPreviewType('live')){
+              return '';
+            } else {
+              return view.model.blueprint.get('title');
+            }
+          },
```
Comment (caseymm) [inline appjs/views/edit_project.js]: We no longer show the title of the blueprint within the project form for blueprints that are live preview enabled.

File: appjs/views/mixins/actions.js [hunk 12]
```diff
@@ -36,6 +36,10 @@ module.exports = {
       $btn.button('loading');
     }
 
+    if(action === 'build-and-publish'){
+      view.$('#projectForm form').submit();
+    }
```
Comment (caseymm) [inline appjs/views/mixins/actions.js]: Since we are not hiding the publish button after the initial save, and since the user may change the share text on the publish tab, we want to make sure that the project is saved before building.

File: appjs/views/mixins/form.js [hunk 13]
```diff
@@ -73,21 +73,31 @@ module.exports = {
...
+            if(model_class === 'Project'){
+              if(view.model.hasUnpublishedUpdates()){
+                app.view.info(model_class+' updates saved but not published');
+              } else {
+                app.view.success(model_class+' updates saved');
+              }
+            } else {
+              app.view.success(model_class+' updates saved');
+            }
+          }
```
Comment (caseymm) [inline appjs/views/mixins/form.js]: Prevents overlapping notification on project save/build. If the project is `publishing`, the alert status will remain as "Publishing..." and not be overridden by "updates saved."

File: appjs/views/mixins/form.js [hunk 14]
```diff
@@ -73,21 +73,31 @@ module.exports = {
...
         logger.debug('next: '+next);
         if ( next === 'show' && action === 'new' ) {
-          var updatedFormData = view.alpaca.getValue();
-          delete updatedFormData['slug'];
-          view.formDataOnLoad = updatedFormData;
+          if(model_class === 'Project'){
+            var updatedFormData = view.alpaca.getValue();
+            delete updatedFormData['slug'];
+            view.formDataOnLoad = updatedFormData;
+          }
```
Comment (caseymm) [inline appjs/views/mixins/form.js]: This fixes an error that appears when adding new blueprints that was created when the save notification was added. The app was looking for `view.alpaca.getValue()` when the blueprint was saved, but since alpaca is not used there, it was coming back as undefined.

File: package.json [hunk 15]
```diff
@@ -32,7 +32,7 @@
     "test": "./bin/run_test_server"
   },
   "dependencies": {
-    "alpaca": "voxmedia/alpaca#dev",
+    "alpaca": "caseymm/alpaca#dev-input-style",
```
Comment (caseymm) [inline package.json]: This will need to be switched before we can merge. I had to make a few tweaks to alpaca to get a few style related things in forms to work correctly and don't have access to the repo. I will submit a PR for this branch of alpaca as well.

File: appjs/views/edit_project.js [hunk 16]
```diff
@@ -170,85 +210,113 @@ var EditProject = BaseView.extend(require('./mixins/actions'), require('./mixins
...
+      $('.project-saved').show().css('display', 'inline-block');
     }
 
     // Make sure the form is valid before proceeding
     // Alpaca takes a loooong time to validate a complex form
     if ( !this.formValidate(this.model, $form) ) {
       // If the form isn't valid, bail
       return;
-    }
+    } else {
```
Comment (caseymm) [inline appjs/views/edit_project.js]: Prevents the project preview from rendering until the form is valid.

File: appjs/views/application.js [hunk 17]
```diff
@@ -135,9 +146,15 @@ var Application = BaseView.extend(require('./mixins/links.js'), {
     }
   },
 
+  toggleNav: function(event){
+    logger.debug('toggleNav', event);
+    if($('#navbar-collapse').hasClass('in')){
+      $('#navbar-collapse').collapse('toggle');
+    }
+  },
```
Comment (caseymm) [inline appjs/views/application.js]: Needed in order to get the mobile nav bar to retract correctly after a page has been selected.

File: appjs/templates/theme_list.ejs [hunk 18]
```diff
@@ -53,59 +59,135 @@
...
+        <% } %>
+
+      </ul>
+
+      <% if(collection.state.currentPage !== collection.state.lastPage){ %>
+        <li class="next<%=hasNextPage() ? '' : ' disabled' %>"><a href="<%=getNextPageUrl()%>">Next</a></li>
+      <% } %>
+    </ul>
+  </nav>
+  <% } %>
```
Comment (caseymm) [inline appjs/templates/theme_list.ejs]: Pagination mirroring the style of Anthem's isn't built into bootstrap (as far as I could tell), so I added this. Happy to change, though, if there's a more succinct way of doing this.

File: app/assets/stylesheets/autotune/_core.scss [hunk 19]
```diff
@@ -0,0 +1,888 @@
+//*\\//*\\//*\\//*\\//*\\
+// Main
+//*\\//*\\//*\\//*\\//*\\
+
+html{
+  overflow-x:hidden;
+  height: auto;
+}
+
+h1, h2, h3, h4, h5, h6, .h1, .h2, .h3, .h4, .h5, .h6{
```
Comment (ryanmark) [inline app/assets/stylesheets/autotune/_core.scss]: Why is this here and not in the fashion stuff? Also why the classes?

File: app/assets/stylesheets/autotune/_core.scss [hunk 20]
```diff
@@ -0,0 +1,888 @@
...
+
+.small-list-view{
+  display: none;
+  @media (max-width: $screen-sm){
+    display: inline;
+  }
+}
+
+//page headings
+.m-page-heading {
```
Comment (ryanmark) [inline app/assets/stylesheets/autotune/_core.scss]: Whats the m-page-heading for?

File: app/assets/stylesheets/autotune/_core.scss (hunk 19 above)
Comment (caseymm) [inline app/assets/stylesheets/autotune/_core.scss]: This section on blueprints and themes.
screen shot 2016-07-06 at 12 11 42 pm

File: appjs/templates/theme_list.ejs (hunk 18 above)
Comment (ryanmark) [inline appjs/templates/theme_list.ejs]: Did you create a whole new set of classes or did you override the bootstrap stuff?

File: app/assets/stylesheets/autotune/_core.scss (hunk 19 above)
Comment (ryanmark) [inline app/assets/stylesheets/autotune/_core.scss]: Why use `.h1` instead of `h1` ?

File: appjs/views/edit_project.js [hunk 21]
```diff
@@ -153,6 +154,45 @@ var EditProject = BaseView.extend(require('./mixins/actions'), require('./mixins
...
+          $('#form-pane').css("width", view.formWidth);
+          $('#preview-pane').css("width", event.pageX);
+          view.showPreviewButtons();
+        }
+      });
+    }
+
+    $(document).mouseup(function (e) {
+      $('#embed-preview').removeClass('screen');
+      $(document).unbind('mousemove');
```
Comment (ryanmark) [inline appjs/views/edit_project.js]: Looks like this has the potential to unbind all mousemove listeners, even ones you don't know about.

File: appjs/templates/theme_list.ejs (hunk 18 above)
Comment (caseymm) [inline appjs/templates/theme_list.ejs]: I created new ones. It slipped my mind that the ones that were in there (I'm assuming?) came from bootstrap. I can change them.

**Comment by ryanmark** [comment]
There is too much here for me to fully wrap my head around today. A few of the things that jumped out at me I commented on above. Lets get those resolved, then 👍 . As I get to know these changes better I'll raise anything that looks problematic.

**Comment by ryanmark** [comment]
I reran the build test and it passed

File: appjs/templates/theme_list.ejs (hunk 18 above)
Comment (caseymm) [inline appjs/templates/theme_list.ejs]: Oh - I take that back, kind of. I got confused as to which comment I was replying to... apparently replying on my phone while on a train on a bridge didn't work out so well. 

I modified the classes that were already there for the most part, but did have add to them a bit; plus, I had to create the extra buttons since bootstrap doesn't do that.
